File: kvstore/proxy/router.py
"""Consistent hash ring with virtual nodes."""

from __future__ import annotations

import hashlib
import threading
from bisect import bisect_left, insort


class HashRing:
    """
    A consistent-hash ring mapping arbitrary string keys to node addresses.

    Virtual nodes are inserted for each physical node to even out the
    distribution and minimise reshuffling when nodes join or leave.
    """

    def __init__(self, nodes: list[str] | None = None, virtual_nodes: int = 100) -> None:
        self._virtual_nodes = virtual_nodes
        self._mu = threading.Lock()
        self._ring: dict[int, str] = {}   # hash → node address
        self._sorted_keys: list[int] = []

        for node in (nodes or []):
            self.add_node(node)

    # ── ring management ──────────────────────────

    def add_node(self, node: str) -> None:
        with self._mu:
            for i in range(self._virtual_nodes):
                key = self._hash(f"{node}#{i}")
                self._ring[key] = node
                insort(self._sorted_keys, key)

    def remove_node(self, node: str) -> None:
        with self._mu:
            for i in range(self._virtual_nodes):
                key = self._hash(f"{node}#{i}")
                if key in self._ring:
                    del self._ring[key]
                    idx = bisect_left(self._sorted_keys, key)
                    if idx < len(self._sorted_keys) and self._sorted_keys[idx] == key:
                        self._sorted_keys.pop(idx)

    # ── key routing ──────────────────────────────

    def get_node(self, key: str) -> str | None:
        with self._mu:
            if not self._ring:
                return None
            h = self._hash(key)
            idx = bisect_left(self._sorted_keys, h)
            if idx == len(self._sorted_keys):
                idx = 0
            return self._ring[self._sorted_keys[idx]]

    def nodes(self) -> list[str]:
        with self._mu:
            return list(dict.fromkeys(self._ring.values()))

    # ── internal ─────────────────────────────────

    @staticmethod
    def _hash(data: str) -> int:
        digest = hashlib.md5(data.encode("utf-8"), usedforsecurity=False).hexdigest()
        return int(digest, 16)

File: proxy/router.go
package proxy

import (
	"crypto/md5"
	"encoding/binary"
	"sort"
	"strconv"
	"sync"
)

// HashRing maps arbitrary keys to node addresses using virtual nodes.
type HashRing struct {
	virtualNodes int
	mu           sync.Mutex
	ring         map[uint64]string
	sortedKeys   []uint64
}

func NewHashRing(nodes []string, virtualNodes int) *HashRing {
	if virtualNodes <= 0 {
		virtualNodes = 100
	}
	r := &HashRing{
		virtualNodes: virtualNodes,
		ring:         make(map[uint64]string),
	}
	for _, node := range nodes {
		r.AddNode(node)
	}
	return r
}

func (r *HashRing) AddNode(node string) {
	r.mu.Lock()
	defer r.mu.Unlock()

	for i := 0; i < r.virtualNodes; i++ {
		h := hash(node + "#" + strconv.Itoa(i))
		if _, exists := r.ring[h]; !exists {
			r.sortedKeys = append(r.sortedKeys, h)
		}
		r.ring[h] = node
	}
	sort.Slice(r.sortedKeys, func(i, j int) bool {
		return r.sortedKeys[i] < r.sortedKeys[j]
	})
}

func (r *HashRing) RemoveNode(node string) {
	r.mu.Lock()
	defer r.mu.Unlock()

	for i := 0; i < r.virtualNodes; i++ {
		h := hash(node + "#" + strconv.Itoa(i))
		if _, exists := r.ring[h]; exists {
			delete(r.ring, h)
			idx := sort.Search(len(r.sortedKeys), func(i int) bool {
				return r.sortedKeys[i] >= h
			})
			if idx < len(r.sortedKeys) && r.sortedKeys[idx] == h {
				r.sortedKeys = append(r.sortedKeys[:idx], r.sortedKeys[idx+1:]...)
			}
		}
	}
}

func (r *HashRing) GetNode(key string) (string, bool) {
	r.mu.Lock()
	defer r.mu.Unlock()

	if len(r.sortedKeys) == 0 {
		return "", false
	}
	h := hash(key)
	idx := sort.Search(len(r.sortedKeys), func(i int) bool {
		return r.sortedKeys[i] >= h
	})
	if idx == len(r.sortedKeys) {
		idx = 0
	}
	return r.ring[r.sortedKeys[idx]], true
}

func (r *HashRing) Nodes() []string {
	r.mu.Lock()
	defer r.mu.Unlock()

	seen := make(map[string]struct{})
	nodes := make([]string, 0)
	for _, key := range r.sortedKeys {
		node := r.ring[key]
		if _, ok := seen[node]; ok {
			continue
		}
		seen[node] = struct{}{}
		nodes = append(nodes, node)
	}
	return nodes
}

func hash(data string) uint64 {
	sum := md5.Sum([]byte(data))
	return binary.BigEndian.Uint64(sum[:8])
}
